File: app/retrieval/context_reader.py
from __future__ import annotations

from collections import defaultdict
from pathlib import Path

from app.schemas.retrieval import Evidence, GrepHit


class ContextReader:
    def __init__(self, knowledge_base_dir: Path):
        self.knowledge_base_dir = knowledge_base_dir

    def expand(self, hits: list[GrepHit], window_size: int = 20) -> list[Evidence]:
        grouped: dict[str, list[GrepHit]] = defaultdict(list)
        for hit in hits:
            grouped[hit.file].append(hit)

        evidence: list[Evidence] = []
        radius = max(1, window_size // 2)
        for file, file_hits in grouped.items():
            path = self.knowledge_base_dir / file
            if not path.exists():
                continue
            lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
            ranges = []
            for hit in file_hits:
                start = max(1, hit.line_number - radius)
                end = min(len(lines), hit.line_number + radius)
                ranges.append((start, end, set(hit.matched_terms)))
            for start, end, terms in self._merge_ranges(ranges):
                text = "\n".join(f"{idx}: {lines[idx - 1]}" for idx in range(start, end + 1))
                evidence.append(
                    Evidence(
                        file=file,
                        line_start=start,
                        line_end=end,
                        score=0.0,
                        matched_terms=sorted(terms),
                        text=text,
                    )
                )
        return evidence

    def _merge_ranges(self, ranges: list[tuple[int, int, set[str]]]) -> list[tuple[int, int, set[str]]]:
        merged: list[tuple[int, int, set[str]]] = []
        for start, end, terms in sorted(ranges):
            if not merged or start > merged[-1][1] + 1:
                merged.append((start, end, set(terms)))
            else:
                old_start, old_end, old_terms = merged[-1]
                merged[-1] = (old_start, max(old_end, end), old_terms | terms)
        return merged

File: src/retrieval/contextReader.ts
import fs from "node:fs/promises";
import path from "node:path";
import { type Evidence, type GrepHit } from "../types.js";

export class ContextReader {
  constructor(private knowledgeBaseDir: string) {}

  async expand(hits: GrepHit[], windowSize = 20): Promise<Evidence[]> {
    const grouped = new Map<string, GrepHit[]>();
    for (const hit of hits) grouped.set(hit.file, [...(grouped.get(hit.file) ?? []), hit]);
    const evidence: Evidence[] = [];
    const radius = Math.max(1, Math.floor(windowSize / 2));
    for (const [file, fileHits] of grouped) {
      const filePath = path.join(this.knowledgeBaseDir, file);
      const lines = (await fs.readFile(filePath, "utf8")).split(/\r?\n/);
      const ranges = fileHits.map((hit) => ({
        start: Math.max(1, hit.lineNumber - radius),
        end: Math.min(lines.length, hit.lineNumber + radius),
        terms: new Set(hit.matchedTerms)
      }));
      for (const range of mergeRanges(ranges)) {
        evidence.push({
          file,
          lineStart: range.start,
          lineEnd: range.end,
          score: 0,
          matchedTerms: [...range.terms].sort(),
          text: Array.from({ length: range.end - range.start + 1 }, (_v, idx) => {
            const lineNo = range.start + idx;
            return `${lineNo}: ${lines[lineNo - 1] ?? ""}`;
          }).join("\n")
        });
      }
    }
    return evidence;
  }
}

function mergeRanges(ranges: Array<{ start: number; end: number; terms: Set<string> }>): Array<{ start: number; end: number; terms: Set<string> }> {
  const merged: Array<{ start: number; end: number; terms: Set<string> }> = [];
  for (const range of ranges.sort((a, b) => a.start - b.start)) {
    const last = merged.at(-1);
    if (!last || range.start > last.end + 1) merged.push({ start: range.start, end: range.end, terms: new Set(range.terms) });
    else {
      last.end = Math.max(last.end, range.end);
      for (const term of range.terms) last.terms.add(term);
    }
  }
  return merged;
}
